Add support for clang-cl on Windows

jon-turney opened the pull request:
This adds support for `clang-cl` (which might be found under the name `cl`) on Windows.

WIP because:
* ~~Adds another appveyor job. (Probably want to do #4154 first)~~
* ~~I've chosen to make `compiler.get_id()` return `msvc` in this case, so existing code which checks this does the right thing.  But anything that checks `compiler.version()` is going to get something unexpected (a clang version number, rather than a msvc version number)...~~
* ~~The change to `CompilerHolder.determine_args()` should be written a better way.~~
* ~~Use `llvm-lib` instead of `lib`?~~

jpakkane commented:
Looks good thus far.

jon-turney commented:
Update for LLVM 7 in the appveyor image.  This only provides `clang-cl` under that name, not also as `cl`.  Make arrangements to use `clang-cl`.  Fix various issues using that name that exposes.  Fix pch and unknown argument handling to deal with differences between `clang-cl`  and `cl`.

jpakkane commented:
Is this still a WIP or is it ready to merge?

If the former, what is still missing.

jon-turney commented:
I think this is in a state where it's a useful incremental change.

There's still some work to be done to follow up:

* Relies on MSVS for `LIB` and `RC`.  Should be able to do something useful when MSVS isn't installed(use llvm-lib, skip windres tests???)
* Some way to distinguish `clang-cl` and `cl` so `compiler.version()` is useful

jpakkane commented:
:crossed_swords: conflicts.

jon-turney commented:
Rebased to fix conflict.  

Fixed `detect_vs_dep_prefix()` to invoke the detected compiler, not hardcode `cl` (which would fail if that isn't in `PATH`)

jon-turney commented:
> Any estimates on how close #4154 is to finished?

It works in the sense that that tests run and pass, but I'm not going to claim it's complete, unchanging and perfect :)

> I would like to get this in but I don't really like to make CI any slower than it already is.

Yeah, I think I probably need to look at moving the test for this PR to azure.

dcbaker commented:
So, about compiler id. I'm working on ICC for windows right now, which also behaves mostly like MSVC with some exceptions. Since we have a bunch of compilers that try to behave like the dominant compiler on a platform, what if we added a two new methods to the compiler object, `is_gcc_like` and `is_msvc_like`? that way a build system can still take appropriate action if the msvc-like compiler is Clang or ICC, but it can actually know the difference between real MSVC and another compiler that behaves a lot like MSVC? This is really critical for ICC because on Windows it behaves like MSVC but on Linux and MacOS it behaves like a mix of GCC and Clang.

jon-turney commented:
> about compiler id.

Yeah, on second thoughts, what I've done with `compiler.get_id()` might be a mistake.

Looking at the uses in meson tests cases, I'd say about half are where we are constructing a compiler invocation by hand for the purposes of the test, and use ` compiler.get_id() == 'msvc'` to determine if it's argument style is cl-like or gcc-like.  So maybe adding something like `compiler.get_arg_type()` which returns `msvc` or `gcc` would be  a better approach...

ignatenkobrain commented on mesonbuild/backend/backends.py:
```diff
@@ -23,7 +23,7 @@
 from ..mesonlib import classify_unity_sources
 from ..mesonlib import File
-from ..compilers import CompilerArgs, get_macos_dylib_install_name
+from ..compilers import CompilerArgs, get_macos_dylib_install_name, VisualStudioCCompiler
```
__[Flake8]__

[F401] '..compilers.get_macos_dylib_install_name' imported but unused

jon-turney commented:
Reworked based on #4378.  Added azure CI job.

jpakkane commented:
⚔️ conflicts.

Does this still need to have an appveyor job as well? It probably slows it down a whole bunch.

jon-turney commented:
> Does this still need to have an appveyor job as well? It probably slows it down a whole bunch.

Nope.  I'll get rid of that.

> I tested this branch on Windows so that `cl.exe` is not in path:
> 
>   * Windows specific tests fail because there is no resource compiler in path
>   * `run_project_tests.py` fails because of `detect_system_compiler`
>   * all tests that use shared libraries fail to run
>   * a bunch of other project tests fail as well

> There's still some work to be done to follow up:
> 
>  * Relies on MSVS for `LIB` and `RC`.  Should be able to do something useful when MSVS isn't installed(use llvm-lib, skip windres tests???)

Yeah, this doesn't work yet.  I have some more changes for that.

I'm not sure if clang-cl can be used with just the Windows SDK (possibly there's some configuration needed to use ucrt rather than msvcrt...)

ignatenkobrain commented on mesonbuild/environment.py:
```diff
@@ -900,14 +917,14 @@ def detect_static_linker(self, compiler):
         popen_exceptions = {}
         for linker in linkers:
-            if 'lib' in linker or 'lib.exe' in linker:
+            if not set(['lib','lib.exe','llvm-lib','llvm-lib.exe']).isdisjoint(linker):
```
__[Flake8]__

[E231] missing whitespace after ','

jon-turney commented:
Remove appveyor job.  Add fixes for not having VC tools `cl`, `lib` or `link` in the path.

Depends on #4378.  Please can we resolve the discussion there before merging this.

> I'm not sure if clang-cl can be used with just the Windows SDK (possibly there's some configuration needed to use ucrt rather than msvcrt...)

So, I was somewhat misremembering the issues here:

* clang-cl automatically uses correct include and lib search paths, without needing LIB & INCLUDE env vars set, but takes notice of them if they are set.
* lld-link doesn't appear to, so needs LIB set correctly
* rc needs INCLUDE set correctly

> Should be able to do something useful when MSVS isn't installed

I'm not sure this is possible currently (or meant to be).  The Windows SDK includes depend on vcruntime.h, and linking ends up requiring libcmt (both from VC).

The best solution I found was to import the env vars used by VC, then remove the PATH changes it makes.  Perhaps @mstorsjo can offer some insight here.

mstorsjo commented:
> > Should be able to do something useful when MSVS isn't installed
> 
> I'm not sure this is possible currently (or meant to be). The Windows SDK includes depend on vcruntime.h, and linking ends up requiring libcmt (both from VC).

I don't think it's supposed to do anything sensible without MSVS installed (or the headers/libs from MSVS extracted and copied to linux). As clang-cl (or `clang -target <arch>-windows-msvc`) emulates MSVC, it really needs those headers from there (from any of the recent MSVC versions). I don't think it's expected to work with only the Windows SDK.

> The best solution I found was to import the env vars used by VC, then remove the PATH changes it makes. Perhaps @mstorsjo can offer some insight here.

That's probably it. Either that, or first run the vcvars bat files for setting up the environment, then adding the directory with clang-cl first in the path.

jon-turney commented:
> > The best solution I found was to import the env vars used by VC, then remove the PATH changes it makes. Perhaps @mstorsjo can offer some insight here.
> 
> That's probably it. Either that, or first run the vcvars bat files for setting up the environment, then adding the directory with clang-cl first in the path.

Yeah, that would be fine for normal usage.  But I think removing `cl` and `link` from the PATH is appropriate for our CI setup, since we seem to like to hardcode those names.

jon-turney commented:
Rebased now #4378  is merged.

amitdo commented:
https://github.com/mesonbuild/meson/blob/68c83cb213ab3f17593699020313535d5de1e92f/docs/markdown/Reference-tables.md#compiler-ids
`clang-cl` is missing in this list.

jon-turney commented:
> `clang-cl` is missing in this list.

Thanks for pointing out that oversight.  Fixed in #4538.

You can easily submit doc changes like this using the 'edit on gitthub' button on http://mesonbuild.com/Reference-tables.html
